utils/kafka-test-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/KafkaClient.scala

```diff
@@ -1,56 +1,136 @@
+    val strContent = content match {
+      case str: String => str
+      case _ => implicitly[Encoder[T]].apply(content).noSpaces
+    }
+
     val promise = Promise[RecordMetadata]()
-    val record = partition.map(new ProducerRecord[String, String](topic, _, key, content)).getOrElse(new ProducerRecord[String, String](topic, key, content))
+    val record = createRecord(topic, key, strContent, partition, timestamp, headers)
     producer.send(record, producerCallback(promise))
     promise.future
   }
 
-  def sendMessage(topic: String, content: String): Future[RecordMetadata] = {
-    val promise = Promise[RecordMetadata]()
-    producer.send(new ProducerRecord[String, String](topic, content), producerCallback(promise))
-    promise.future
+  def consumeLastMessage[T: Decoder](topic: String): KeyMessage[String, T] =
+    consumeMessages[T](topic, LastMessageCount).head
+
+  def consumeMessages[V: Decoder](topic: String, count: Int, shouldSeekToBeginning: Boolean = false): List[KeyMessage[String, V]] =
+    consumeRawMessages(topic, count, shouldSeekToBeginning).map { record =>
+      KeyMessage(record.keyAsStr, record.valueAs[V], record.timestamp)
+    }
+
+  def consumeLastStrMessage(topic: String): KeyMessage[String, String] =
+    consumeStrMessages(topic, LastMessageCount).head
+
+  def consumeStrMessages(topic: String, count: Int, shouldSeekToBeginning: Boolean = false): List[KeyMessage[String, String]] =
+    consumeRawMessages(topic, count, shouldSeekToBeginning).map { record =>
+      KeyMessage(record.keyAsStr, record.valueAsStr, record.timestamp)
+    }
+
+  def consumeLastRawMessage(topic: String, shouldSeekToBeginning: Boolean = false): ConsumerRecord[Array[Byte], Array[Byte]] =
+    consumeRawMessages(topic, LastMessageCount, shouldSeekToBeginning).head
+
+  def consumeRawMessages(topic: String, count: Int, shouldSeekToBeginning: Boolean = false): List[ConsumerRecord[Array[Byte], Array[Byte]]] = {
+    (consumers.get(topic) match {
+      case Some(consumer) =>
+        if (shouldSeekToBeginning) {
+          seekToBeginning(topic, consumer)
+        }
+        consumer
+      case None =>
+        createConsumer(topic)
+    }).consumeWithConsumerRecord(topic)
+      .take(count)
+      .toList
   }
 
+  private def createConsumer(topic: String, groupIdOpt: Option[String] = None): KafkaConsumer[Array[Byte], Array[Byte]] = synchronized {
+    // each consumer by default is in other consumer group to make sure that messages won't be stolen by other consumer consuming the same topic
+    val groupId = groupIdOpt.getOrElse(s"testGroup_${UUID.randomUUID()}")
+    val props = KafkaTestUtils.createConsumerConnectorProperties(kafkaAddress, groupId)
+    val consumer = new KafkaConsumer[Array[Byte], Array[Byte]](props)
+    consumers += topic -> consumer
```

Comment:
What about a consumer that reads messages from multiple topics? What do you need to keep consumers in a Map with topic as a key?

Reply:
How many tests in that situation do you have? This KafkaClient is used only for tests, this is optimization for creating new consumer which is pretty costly...